packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt

```diff
@@ -512,15 +510,135 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
+            val fieldSerializedData = serializedModelData[key]
+
+            // if the field serialized value is null
+            // assign null to this field in the deserialized map
+            if (fieldSerializedData == null) {
+                result[key] = fieldSerializedData
+                continue
+            }
+
+            if (field.isModel) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldModelSchema = modelProvider.modelSchemas()[field.targetType] ?: continue
+                result[key] = SerializedModel.builder()
+                    .serializedData(deserializeNestedModel(fieldSerializedData as Map<String, Any?>, fieldModelSchema))
+                    .modelSchema(fieldModelSchema)
                     .build()
-            } else
-                it.value
+            } else if (field.isCustomType) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldCustomTypeSchema = modelProvider.customTypeSchemas()[field.targetType] ?: continue
+
+                // When a field is custom type
+                // the field value can only be a single custom type
+                // or a list of items of the same custom type
+                if (field.isArray) {
+                    result[key] = (fieldSerializedData as List<Map<String, Any?>>).map { listItem ->
+                        SerializedCustomType.builder()
+                            .serializedData(deserializeNestedCustomType(listItem, fieldCustomTypeSchema))
+                            .customTypeSchema(fieldCustomTypeSchema)
+                            .build()
+                    }
+                } else {
+                    result[key] = SerializedCustomType.builder()
+                        .serializedData(
+                            deserializeNestedCustomType(fieldSerializedData as Map<String, Any>, fieldCustomTypeSchema)
+                        )
+                        .customTypeSchema(fieldCustomTypeSchema)
+                        .build()
+                }
+            } else {
+                result[key] = fieldSerializedData
+            }
+        }
+
+        return result.toMap()
+    }
+
+    private fun deserializeNestedCustomType(
+        serializedModelData: Map<String, Any?>, customTypeSchema: CustomTypeSchema): Map<String, Any?> {
+        val result = mutableMapOf<String, Any?>()
+
+        for ((key, field) in customTypeSchema.fields.entries) {
+            if (!serializedModelData.containsKey(key)) {
+                continue
+            }
+
+            val fieldSerializedData = serializedModelData[key]
+
+            // if the filed serialized value is null
+            // assign null to this field in the deserialized map
+            if (fieldSerializedData == null) {
+                result[key] = fieldSerializedData
+                continue
+            }
+
+            if (field.isCustomType) {
```

Comment:
Can custom types have a field with a model type? For example, is this a valid schema?

```graphql
type Blog @model {
  id: ID!
  name: String!
  authorData: AuthorData!
}

type AuthorData {
  primaryAuthor: User!
  secondaryAuthor: User
}

type User @model {
  id: ID!
  name: String!
}
```

Reply:
This is not supported, as CustomType is underlying a single field of a model, a part of a single field's value is not able to created model relationship. AdminUI data modeling doesn't provide "add a relationship" functionality under a CustomType.